handle spotify web URLs as well

squeed opened the pull request:
This automatically converts spotify web URLs (e.g. https://open.spotify.com/track/XXX) to spotify URIs.

This is particularly useful to me, since I have a NFC-tag jukebox for the kids to use. And it's much easier to write a standard URL to an NFC tag rather than changing it to a spotify URI.

Thanks!

squeed commented:
Thanks for kick-starting CI. That doesn't seem like an issue caused by this PR; should I take a look?

squeed commented:
@fondberg I've rebased on latest master.

Leicas commented:
I just updated to fix the device_id casting issue to the latest branch and did not realize I needed this for the NFC-tag jukebox I also use for my kids. Any way I can help integrating this for a merge in master ?

squeed commented:
@Leicas I worked around this with the following action template:

```yaml
service: spotcast.start
data:
  entity_id: media_player.wohnzimmer
  uri: |-
    {% set url = trigger.event.data.url %} {% if url.startswith('http') %}
      {% set parts = url.split('/') %}
      spotify:{{ parts[3] }}:{{ parts[4].split('?')[0] }}
    {% else %}
      {{ url }}
    {% endif %}
```

fcusson commented on custom_components/spotcast/helpers.py:
```diff
@@ -171,6 +171,28 @@ def get_random_playlist_from_category(spotify_client:spotipy.Spotify, category:s
     return chosen['uri']
 
+
+def url_to_spotify_uri(url: str) -> str:
```
do not return None on error, Raise an error instead to let the parent process what to do on error

squeed replied:
Sure, makes sense. What's the convention around error types for the project?

fcusson commented on custom_components/spotcast/__init__.py:
```diff
@@ -206,6 +207,9 @@ def start_casting(call: ha_core.ServiceCall):
             # remove ? from badly formatted URI
             uri = uri.split("?")[0]
 
+            if uri.startswith("https"):
```
raise logger error on invalid url. This should be a specific error message different than the URI. The current process could lead people to believe URL are not trully compatible and force them to revert to URI, instead of fixing the url

fcusson reviewed (changes requested):
Sorry for the late reply on this one. Small changes requested, this is mostly to ensure your codes fit in the future rewrite I am working on which, one of its part is improving logging and error management.

squeed commented:
Updated, PTAL. Thanks!

fcusson reviewed (approved):
LGTM
